src/vm/circuit.rs

```diff
@@ -0,0 +1,333 @@
+            eval_u96_limbs_less_than_guarantee_verify(registry, info, args)
+        }
+        CircuitConcreteLibfunc::U96SingleLimbLessThanGuaranteeVerify(info) => {
+            eval_u96_single_limb_less_than_guarantee_verify(registry, info, args)
+        }
+    }
+}
+
+pub fn eval_add_input(
+    _registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    _info: &SignatureAndTypeConcreteLibfunc,
+    args: Vec<Value>,
+) -> EvalAction {
+    let [Value::Circuit(mut values), Value::Struct(members)]: [Value; 2] = args.try_into().unwrap()
+    else {
+        panic!()
+    };
+    assert_ne!(values.len(), values.capacity());
+
+    let [Value::U128(l0), Value::U128(l1), Value::U128(l2), Value::U128(l3)]: [Value; 4] =
+        members.try_into().unwrap()
+    else {
+        panic!()
+    };
+
+    let l0 = l0.to_le_bytes();
+    let l1 = l1.to_le_bytes();
+    let l2 = l2.to_le_bytes();
+    let l3 = l3.to_le_bytes();
+    values.push(BigUint::from_bytes_le(&[
+        l0[0], l0[1], l0[2], l0[3], l0[4], l0[5], l0[6], l0[7], l0[8], l0[9], l0[10],
+        l0[11], //
+        l1[0], l1[1], l1[2], l1[3], l1[4], l1[5], l1[6], l1[7], l1[8], l1[9], l1[10],
+        l1[11], //
+        l2[0], l2[1], l2[2], l2[3], l2[4], l2[5], l2[6], l2[7], l2[8], l2[9], l2[10],
+        l2[11], //
+        l3[0], l3[1], l3[2], l3[3], l3[4], l3[5], l3[6], l3[7], l3[8], l3[9], l3[10],
+        l3[11], //
+    ]));
+
+    EvalAction::NormalBranch(
+        (values.len() != values.capacity()) as usize,
+        smallvec![Value::Circuit(values)],
+    )
+}
+
+pub fn eval_eval(
+    _registry: &ProgramRegistry<CoreType, CoreLibfunc>,
+    _info: &SignatureAndTypeConcreteLibfunc,
+    _args: Vec<Value>,
+) -> EvalAction {
+    let [add_mod @ Value::Unit, mul_mod @ Value::Unit, _descripctor @ Value::Unit, Value::Circuit(inputs), Value::CircuitModulus(modulus), _, _]: [Value; 7] = _args.try_into().unwrap()
+    else {
+        panic!()
+    };
+    let circ_info = match _registry.get_type(&_info.ty).unwrap() {
+        CoreTypeConcrete::Circuit(CircuitTypeConcrete::Circuit(info)) => &info.circuit_info,
+        _ => todo!(),
+    };
+    let mut outputs = vec![None; 1 + circ_info.n_inputs + circ_info.values.len()];
+    let mut add_gates = circ_info.add_offsets.iter().peekable();
+    let mut mul_gates = circ_info.mul_offsets.iter().skip(circ_info.n_inputs);
+
+    outputs[0] = Some(BigUint::from(1_u8));
+
+    for (i, input) in inputs.iter().enumerate() {
+        let gate = &circ_info.mul_offsets[i];
+        outputs[gate.output] = Some(input.to_owned());
+    }
+
+    let success = loop {
+        while let Some(add_gate) = add_gates.peek() {
+            let lhs = outputs[add_gate.lhs].to_owned();
+            let rhs = outputs[add_gate.rhs].to_owned();
+
+            match (lhs, rhs) {
+                (Some(l), Some(r)) => {
+                    outputs.insert(add_gate.output, Some((l + r) % &modulus));
```

Comment:
You should use indexing (`outputs[add_gate.output] = Some(...)`) rather than calling the `insert` method. The insert method actually adds a new element to the array (increasing its length and shifting all the elements in the right)